### ISSUE TITLE
[Q] Add one empty line below prompt

### ISSUE THREAD

**Issue by silverqx** [romkatv/powerlevel10k#1457]
Hi, I have to ask, because I can not accomplish this myself, I tried hard a few hours but without success. 👀

I want to achieve to have my command prompt always stuck at the bottom of the terminal and to have one empty line after the prompt, so there will be always one empty line at the bottom of the terminal. 🚀

Or you can look at it that I want to have a prompt that will have 3 lines and the segment on that third line will be empty, so create a new empty segment and put it on the third line of prompt is also way.

This is my current `powerlevel` configuration:
```
DEFAULT_USER="silverqx"
POWERLEVEL9K_STATUS_OK_BACKGROUND="238"

POWERLEVEL9K_LEFT_PROMPT_ELEMENTS=(
  context
  dir
  vcs
)
POWERLEVEL9K_PROMPT_ON_NEWLINE=true
POWERLEVEL9K_PROMPT_ADD_NEWLINE=true
POWERLEVEL9K_MULTILINE_FIRST_PROMPT_PREFIX=""
POWERLEVEL9K_MULTILINE_LAST_PROMPT_PREFIX="λ "
POWERLEVEL9K_RIGHT_PROMPT_ELEMENTS=(
  status_joined
  command_execution_time
  root_indicator
  background_jobs
  history
  node_version
)

POWERLEVEL9K_TRANSIENT_PROMPT=always
# by default is 1, so there is 1 column free before scrollbar
ZLE_RPROMPT_INDENT=0
# stick command prompt at the bottom at start
printf '\n%.0s' {1..$((`tput lines` - 2))}

POWERLEVEL9K_INSTANT_PROMPT=quiet
POWERLEVEL9K_LEGACY_ICON_SPACING=true
```

Thank you for any advice and suggestions. 🙏

**Comment by romkatv** [comment]
Why do you want this?

What do you want to happen if you enter a multi line command (e.g. with ESC ENTER)?

**Comment by silverqx** [comment]
As you can see from my configuration I'm using a transient prompt and I'm already using the terminal that way that the prompt is always at the bottom of the window/terminal and the terminal window is always fullscreen, the reason why I want this is that I want to have a little margin at the bottom.

**Comment by romkatv** [comment]
Thanks. Could you answer my second question?

**Comment by silverqx** [comment]
> Thanks. Could you answer my second question?

Ohhh, I see it now 😁 let me think few mins about it.

**Comment by silverqx** [comment]
> What do you want to happen if you enter a multi line command (e.g. with ESC ENTER)?

@romkatv I got it 😁, when I input multiline command, the empty line should be below the cursor, so it should look like this, the `echo xyz` and `echo another` is some example output for demonstration:
```
> echo xyz
xyz
> echo another
another

~/tmp | git_segment 
λ echo a;
echo b;
echo c;|cursor|
and here I want to have an empty line
```

After I send that multiline command, it should look like this:
```
> echo xyz
xyz
> echo another
another
> echo a;
echo b;
echo c;
a
b
c

~/tmp | git_segment 
λ |cursor|
and here I want to have an empty line
```

**Comment by romkatv** [comment]
Thanks, I think I understand what you want. It's possible to implement.

Firstly, having prompt always at the bottom is a great idea. This is the same setup I use. However, you cannot get this with just powerlevel10k. Many things push prompt upward and then it takes a bunch of <kbd>Enter</kbd> presses to get it down again. The most common offenders here are the completion menu and <kbd>Ctrl-L</kbd>. In my setup this problem is solved by showing the completion menu above cursor and by replacing the builtin `clear-screen` widget. Here's how it looks: 

[![asciicast](https://asciinema.org/a/372068.svg)](https://asciinema.org/a/372068)

You can try this setup here: https://github.com/romkatv/zsh4humans. It provides a few other things apart from keeping prompt always at the bottom of the screen.

Now, to have an empty line at the bottom of the screen you'll need a few things. Firstly, update powerlevel10k to the latest version.

Add this to your config:

```zsh
function precmd() {
  zle && return
  [[ -t 1 ]] || return
  local -i h=2
  print -rn -- ${(pl:$h::\n:)}$'\e['${h}A
}
```

Remove your `printf`.

If you aren't using zsh4humans, add this at the top of `~/.zshrc`:

```zsh
() {
  local esc x y
  IFS='[;' read -s -d R esc\?$'\e[6n' y x <$TTY
  local -i h=2
  print -rn -- ${(pl:$((LINES - y + h))::\n:)}$'\e['${h}A
}
```

Finally, add these parameters:

```zsh
POWERLEVEL9K_NEW_TTY_MAX_AGE_SECONDS=0
POWERLEVEL9K_INSTANT_PROMPT_COMMAND_LINES=1
```

This will print prompt at the right place but if you enter a multi-line command your cursor will end up at the very bottom. This can be solved by wrapping all zle widgets that can modify the buffer and by always maintaining `\n` at the end of it. This can be difficult to do if you aren't experienced with zle. Even if you do this, many things will still place the cursor at the bottom. E.g., pressing ESC X cannot be hooked. In addition, external commands will not respect your desire to keep the bottom line empty. It's a loosing battle. A better approach would be to reposition your terminal or to configure it to have padding at the bottom.

**Comment by silverqx** [comment]
@romkatv Thx 🙏 luxury response, I didn't expect this quality 😎, you answered me all the questions.

Anyway, you have to give me some time to investigate stuffs in your response because they are not everyday things.

`CTRL-L` trick looks great I'm going to reuse it.

I didn't know about `zsh4humans`, I'll dig into it for a while and will see what I can use at my side.

> Remove your print.

Thank you, I see that you enhanced my simple `printf` command with something better, firstly I will have to understand what it is doing and use it.

> If you aren't using zsh4humans, add this at the top of ~/.zshrc

I don't use `zsh4humans`, I will have to decrypt this second code snippet too, looks like that it handles more scenarios than the previous one.

> This can be solved by wrapping all zle widgets that can modify the buffer and by always maintaining \n .... Even if you do this, many things will still place the cursor at the bottom ... It's a loosing battle. 

I have realized that yesterday quickly and I had a strong feeling that it will be so.

As you wrote, the best will be to solve it at the window manager level and simply add padding below the terminal widget and bottom edge, I have already found a solution for `KDE Konsole` that I'm using.

I'm not done yet, but I played yesterday with this, all KDE apps allow to overwrite GUI content with CSS stylesheets and it works really good:
```
QTabWidget::pane {
    padding: 0 0 30px 0;
}
```

Thank you for your time 🙏, when I will have any questions I post it here and I think we can close this.

**Comment by romkatv** [comment]
> I'm not done yet, but I played yesterday with this, all KDE apps allow to overwrite GUI content with CSS stylesheets and it works really good

Great! This is definitely better than hacks within zsh. Now you only need to ensure that prompt starts at the bottom and stays there. You can try zsh4humans for that.

**Comment by silverqx** [comment]
@romkatv What these two settings do?

```
POWERLEVEL9K_NEW_TTY_MAX_AGE_SECONDS=0
POWERLEVEL9K_INSTANT_PROMPT_COMMAND_LINES=1
```

I have decided to use to add padding to the bottom, it will be better than hack zsh, will I need the above settings?

**Comment by romkatv** [comment]
If you are using zsh4humans, you are all set, nothing to do. If you aren't using zsh4humans, set `POWERLEVEL9K_NEW_TTY_MAX_AGE_SECONDS=0` and add the following snippet at the top of `~/.zshrc`:

```zsh
() {
  local esc x y
  IFS='[;' read -s -d R esc\?$'\e[6n' y x <$TTY
  local -i n=$((LINES - y - 2))
  (( n <= 0 )) || print -rn -- ${(pl:$n::\n:)}
}
```

You might want to adjust the value of `n` depending on how tall your prompt is (this code assumes it's 3 lines).

Either way you need to remove that `printf`. The snippet I posted above is better than the `printf` because it won't scroll your terminal window if you run `zsh` within an existing session. It's also faster.

`POWERLEVEL9K_NEW_TTY_MAX_AGE_SECONDS=0` makes the first prompt of a session always have an empty line. This makes it easier to print the right number of empty lines on startup.

**Comment by silverqx** [comment]
@romkatv First I want to thank you 🙏, I have a prompt at the bottom almost all the time now, clearing the screen on the first startup works great and I also reused `-z4h-get-cursor-pos`, `z4h-clear-screen-hard-bottom` and `z4h-clear-screen-soft-bottom` from `zsh4humans` project and assigned them to appropriate shortcuts.

But I have a problem with this in my `.zshrc`, it causes `10sec` startup time of `mc` (midnight commander):
```
() {
  local esc x y
  IFS='[;' read -s -d R esc\?$'\e[6n' y x <$TTY
  local -i n=$((LINES - y - 2))
  (( n <= 0 )) || print -rn -- ${(pl:$n::\n:)}
}
```

Currently, I solved it by adding `[ "$MC_SID" ] && return` at the beginning of the function, it looks like this:
```
() {
  [ "$MC_SID" ] && return
  local esc x y
  IFS='[;' read -s -d R esc\?$'\e[6n' y x <$TTY
  local -i n=$((LINES - y))
  (( n <= 0 )) || print -rn -- ${(pl:$n::\n:)}
}
```

But the same code doesn't cause this problem in your `zsh4humans` project, how did you solve that, do you have any idea what is up? 🤔

**Comment by romkatv** [comment]
From the top of my head I don't know what could cause the 10 second delay.

**Comment by t-dub** [comment]
So I had a slightly different use-case, but since I landed here while searching around, I'll share it.  My scenario is that I have a colorful, rich tmux status line via oh my tmux and a colorful, rich prompt via p10k.  Since I'm using the "always tmux" option in zsh4humans, I'll (almost) never not have both .  I wanted to get some visual separation between the two because I find this hard to parse:

![image](https://user-images.githubusercontent.com/302158/226191278-23fdce62-0004-4e52-916d-d773e760a182.png)


My initial thought was exactly this: I'll just throw a 3rd, empty line in my prompt and off I go.  As outlined here, that does not work quite as cleanly as one would like.

However, I ultimately decided that what would work for me would be to enable pane borders in tmux.  I actually _also_ wanted more visual indication as to which pane I was in, so this solved both problems at once.  In `.tmux.conf` add the following to give you a "status line" which is integrated into the border and then make it empty, which just gives a solid border.  This created the visual separation I was looking for between my prompt and my tmux status line.

```
set -g pane-border-status bottom
set -g pane-border-format ''
```

Much better!
![image](https://user-images.githubusercontent.com/302158/226191301-22323405-88ef-44c8-9ae6-dc911695c241.png)
